src/dawn/SIR/SIR.proto

```diff
@@ -0,0 +1,627 @@
+  bool merge_stages = 2;      // Merge the Stages of this stencil
+  bool merge_do_methods = 3;  // Merge the Do-Methods of this stencil
+  bool merge_temporaries = 4; // Merge the temporaries of this stencil
+  bool use_k_caches = 5;      // Use K-Caches
+}
+
+// @brief Description of a field argument of a Stencil or StencilFunction
+// @ingroup sir_proto
+message Field {
+  string name = 1;        // Name of the field
+  SourceLocation loc = 2; // Source location
+  bool is_temporary = 3;  // Is the field a temporary?
+}
+
+// @brief Directional argument of a StencilFunction
+//
+// Note that this message is merely a placeholder as the actual arguments is encoded as an AST node.
+// @ingroup sir_proto
+message Direction {
+  string name = 1;        // Name of the directional argument
+  SourceLocation loc = 2; // Source location
+}
+
+// @brief Offset argument of a StencilFunction
+//
+// Note that this message is merely a placeholder as the actual arguments is encoded as an AST node.
+// @ingroup sir_proto
+message Offset {
+  string name = 1;        // Name of the offset argument
+  SourceLocation loc = 2; // Source location
+}
+
+// @brief Directional argument of a StencilFunction
+// @ingroup sir_proto
+message StencilFunctionArg {
+  // One of Field, Direction or Offset
+  oneof Arg {
+    Field field_value = 1;
+    Direction direction_value = 2;
+    Offset offset_value = 3;
+  }
+}
+
+// @brief Representation of a vertical interval, given by a lower and upper bound where a bound
+// is represented by a level and an offset (`bound = level + offset`)
+//
+// The Interval has to satisfy the following invariants:
+//  - `lower_level >= Interval::Start`
+//  - `upper_level <= Interval::End`
+//  - `(lower_level + lower_offset) <= (upper_level + upper_offset)`
+//
+// @ingroup sir_proto
+message Interval {
+  // Indication of a special level
+  enum SpecialLevel {
+    Start = 0; // The first level
+    End = 1;   // The last level
+  }
+
+  // Lower level
+  //
+  // A level can be any integer between `[Start, End]` while the `Start` and `End` levels are
+  // treated specially.
+  oneof LowerLevel {
+    SpecialLevel special_lower_level = 1;
+    int32 lower_level = 2;
+  }
+
+  // Upper level
+  //
+  // A level can be any integer between `[Start, End]` while the `Start` and `End` levels are
+  // treated specially.
+  oneof UpperLevel {
+    SpecialLevel special_upper_level = 3;
+    int32 upper_level = 4;
+  }
+
+  int32 lower_offset = 5; // Lower offset added to `lower_level`
+  int32 upper_offset = 6; // Upper offset added to `upper_level`
+}
+
+// @brief Supported builtin types
+// @ingroup sir_proto
+message BuiltinType {
+  enum TypeID { None = 0; Auto = 1; Boolean = 2; Integer = 3; Float = 4; }
+
+  TypeID type_id = 1; // Type identifier
+}
+
+// @brief Supported builtin types
+// @ingroup sir_proto
+message Dimension {
+  enum Direction { I = 0; J = 1; K = 2; Invalid = 3; }
```

Comment:
even if gtclang does not support this, we need to generalize the dimensionality in the SIR.  Keep the one as a canonical dimension and add a custom
```
message CustomDimension {
  string name = 1;
  int32 index;
}
message Dimension {
  anyof dim { CanonicalDimension ;
     CustomDimension;
  };
```